### PR TITLE
automountmap: Fix argument handling.

### PR DESCRIPTION
**Pull request by rjeffman**
Ensure a failure is raised when map name is not provided for arguments,
and allows an empty description to be used to clear automount map
description.

Depends on PR #780

### CODE REVIEW

**Comment by rjeffman** [comment]
Related to https://bugzilla.redhat.com/show_bug.cgi?id=2048552

**Review by varunmylaraiah — APPROVED** [review]
Thanks for the PR.
Changes LGTM.

Console output:
```
[root@master ~]# ipa automountmap-add newlocation newmap --desc="testing empty descreption"
----------------------------
Added automount map "newmap"
----------------------------
  Map: newmap
  Description: testing empty descreption
```


```
- name: Playbook to ensure automount map description is updated with empty.
  hosts: ipaserver
  become: yes
  tasks:
  - ipaautomountmap:
      ipaadmin_principal: admin
      ipaadmin_password: <xxxxx>
      automountlocation: newlocation
      name: newmap
      description: ""

[root@ansible ~]# ansible-playbook -vv -i inventory/server.hosts emptydesc.yaml
ansible-playbook [core 2.12.1]
  config file = /root/ansible.cfg
  configured module search path = ['/root/ansible-freeipa/plugins/modules']
  ansible python module location = /usr/lib/python3.8/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible-playbook
  python version = 3.8.12 (default, Sep 16 2021, 10:46:05) [GCC 8.5.0 20210514 (Red Hat 8.5.0-3)]
  jinja version = 2.10.3
  libyaml = True
Using /root/ansible.cfg as config file
Skipping callback 'default', as we already have a stdout callback.
Skipping callback 'minimal', as we already have a stdout callback.
Skipping callback 'oneline', as we already have a stdout callback.

PLAYBOOK: emptydesc.yaml ****************************************************************************************
1 plays in emptydesc.yaml

PLAY [Playbook to ensure automount map description is updated with empty.] **************************************

TASK [Gathering Facts] ******************************************************************************************
task path: /root/emptydesc.yaml:2
ok: [master.ipadomain.test]
META: ran handlers

TASK [ipaautomountmap] ******************************************************************************************
task path: /root/emptydesc.yaml:7
changed: [master.ipadomain.test] => {"changed": true}
META: ran handlers
META: ran handlers

PLAY RECAP ******************************************************************************************************
master.ipadomain.test      : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

```
[root@master ~]# ipa automountmap-show newlocation newmap
  Map: newmap
[root@master ~]#
```

File: plugins/modules/ipaautomountmap.py [hunk 1]
```diff
@@ -124,9 +127,10 @@ def check_ipa_params(self):
 
     def get_args(self, mapname, desc):  # pylint: disable=no-self-use
         _args = {}
-        if mapname:
-            _args["automountmapname"] = mapname
-        if desc:
+        if not mapname:
+            raise ValueError("Automount map name is required.")
```
Comment (t-woerner) [inline plugins/modules/ipaautomountmap.py:131]: Why is this needed here? A check to have valid names should be done early before any IPA command is used.

File: plugins/modules/ipaautomountmap.py [hunk 2]
```diff
@@ -104,7 +104,10 @@ def get_automountmap(self, location, name):
         except Exception:  # pylint: disable=broad-except
             return None
         else:
-            return response["result"]
+            _result = response["result"]
+            if 'description' not in _result:
+                _result['description'] = ''
+            return _result
```
Comment (t-woerner) [inline plugins/modules/ipaautomountmap.py:110]: Do we have the description issue in more modules? It might more sense to handle a comparison of the empty string in comapre_args_ipa. What do you think?

File: plugins/modules/ipaautomountmap.py (hunk 1 above)
Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py:131]: This is a somewhat paranoid test, but it might happen. Consider this playbook task:
```yaml
- ipaautomountmap:
    name:
      - ''
    description: "an empty name map"
```
This task will pass validation (`len(name) == 1`) on `automountmap` and on (maybe) all other modules, but the name itself is invalid.

Should we test for empty names (inside `for name in names`) or should we rely on IPA API checks?

If we check, there will be less API calls and we can fail early, if we rely on IPA API, we'll have one less error check to deal on modules, but we might get less descriptive errors (I haven't look at the possible errors, I'm just speculating.)

Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py:131]: The issue also might happen on `state: absent`:
```yaml
ipaautomountmap:
  name:
     - test_map
     - ''
     - another_map
   state: absent
```

Comment (t-woerner) [reply plugins/modules/ipaautomountmap.py:131]: I think we should check this before doing anything with the API.

Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py:131]: So... I need we need to fix it on all modules. For `ipagroup`:
```yaml
  - ipagroup:
      name:
        - ""
      state: present
```
This is the result:
```
TASK [ipagroup] ****************************************************************************************
fatal: [fedsrv.ipa.test]: FAILED! => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python3"}, "changed": false, "msg": "There is more than one group ''"}
```
This happens as `ipagroup` still uses `*_find` instead of `*_show`.

Comment (t-woerner) [reply plugins/modules/ipaautomountmap.py:131]: We could add a check to module_params_get. Is there a use case to return a list of strings, with one or more empty strings?

Comment (t-woerner) [reply plugins/modules/ipaautomountmap.py:131]: Please have a look at https://github.com/freeipa/ansible-freeipa/pull/779

Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py:131]: On a fast, first look, #779 looks good (I'm reviewing it). I don't remember of any case were an empty string should be allowed in a list.

Comment (t-woerner) [reply plugins/modules/ipaautomountmap.py:131]: We have one place: userauthtype in the user module.

File: plugins/modules/ipaautomountmap.py (hunk 2 above)
Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py]: I haven't seen this happen in different modules, but I agree that this should be fixed in `compare_args_ipa`.

**Comment by rjeffman** [comment]
This PR should not be merged before #780.

File: plugins/modules/ipaautomountmap.py (hunk 2 above)
Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py]: Comparison has been changed to `module_utils.ansible_freeipa_module.compare_args_ipa`.

File: plugins/modules/ipaautomountmap.py (hunk 1 above)
Comment (rjeffman) [reply plugins/modules/ipaautomountmap.py]: PR #779 has beem merged. Marking this conversation as resolved.

**Comment by t-woerner** [comment]
The changes are part of #780
